x-pack/plugins/security_solution/server/endpoint/lib/artifacts/manifest.ts

```diff
@@ -20,21 +20,46 @@ import {
 } from '../../../../common/endpoint/schema/common';
 import { manifestSchema, ManifestSchema } from '../../../../common/endpoint/schema/manifest';
 import { ManifestEntry } from './manifest_entry';
-import { maybeCompressArtifact, isCompressed } from './lists';
 import { getArtifactId } from './common';
 import { ManifestVersion, manifestVersion } from '../../schemas/artifacts/manifest';
 
+function createInternalManifestEntries(
+  artifactIds: string[],
+  policyId?: string
+): InternalManifestEntrySchema[] {
+  return artifactIds.map((artifactId) => ({ policyId, artifactId }));
+}
+
 export interface ManifestDiff {
-  type: string;
-  id: string;
+  additions: InternalArtifactSchema[];
+  removals: InternalArtifactSchema[];
+  transitions: InternalArtifactSchema[];
+}
+
+export function isEmptyManifestDiff(diff: ManifestDiff) {
+  return diff.additions.length === 0 && diff.removals.length === 0 && diff.transitions.length === 0;
+}
+
+interface ManifestEntryDescriptor {
+  isDefaultEntry: boolean;
+  specificTargetPolicies: Set<string>;
+  entry: ManifestEntry;
+}
+
+function addValueToSet<T>(set?: Set<T>, value?: T) {
+  return new Set([...(set?.values() || []), ...(value !== undefined ? [value] : [])]);
 }
 
 export class Manifest {
-  private entries: Record<string, ManifestEntry>;
+  private readonly allEntries: Map<string, ManifestEntryDescriptor>;
```

Comment:
The need for ManifestEntryDescriptor is probably a stretch, it could probably be done in a simpler way (like separate map to maintain reverse policy to artifact linkage), but it's purely internal implementation detail that is never leaking outside so can be changed afterwards.